wrappers/services.go

```diff
@@ -471,116 +471,180 @@ type EnsureSnapServicesOptions struct {
+}
 
-	defer func() {
-		if err == nil {
-			return
-		}
-		for file, state := range modifiedUnitsPreviousState {
-			if state == nil {
-				// we don't have anything to rollback to, so just remove the
-				// file
-				if e := os.Remove(file); e != nil {
-					inter.Notify(fmt.Sprintf("while trying to remove %s due to previous failure: %v", file, e))
-				}
-			} else {
-				// rollback the file to the previous state
-				if e := osutil.EnsureFileState(file, state); e != nil {
-					inter.Notify(fmt.Sprintf("while trying to rollback %s due to previous failure: %v", file, e))
-				}
+// restore is a helper function which should be called in case any errors happen
+// during the write/update of systemd files
+func (es ensureSnapServicesContext) restore() {
+	for file, state := range es.modifiedUnits {
+		if state == nil {
+			// we don't have anything to rollback to, so just remove the
+			// file
+			if err := os.Remove(file); err != nil {
+				es.inter.Notify(fmt.Sprintf("while trying to remove %s due to previous failure: %v", file, err))
+			}
+		} else {
+			// rollback the file to the previous state
+			if err := osutil.EnsureFileState(file, state); err != nil {
+				es.inter.Notify(fmt.Sprintf("while trying to rollback %s due to previous failure: %v", file, err))
 			}
 		}
-		if modifiedSystem && !preseeding {
-			if e := sysd.DaemonReload(); e != nil {
-				inter.Notify(fmt.Sprintf("while trying to perform systemd daemon-reload due to previous failure: %v", e))
+	}
+
+	if !es.opts.Preseeding {
+		if es.modifiedSystemServices {
+			if err := es.sysd.DaemonReload(); err != nil {
+				es.inter.Notify(fmt.Sprintf("while trying to perform systemd daemon-reload due to previous failure: %v", err))
 			}
 		}
-		if modifiedUser && !preseeding {
-			if e := userDaemonReload(); e != nil {
-				inter.Notify(fmt.Sprintf("while trying to perform user systemd daemon-reload due to previous failure: %v", e))
+		if es.modifiedUserServices {
+			if err := userDaemonReload(); err != nil {
+				es.inter.Notify(fmt.Sprintf("while trying to perform user systemd daemon-reload due to previous failure: %v", err))
 			}
 		}
-	}()
+	}
+}
+
+// reloadModified uses the modifiedSystemServices/modifiedUserServices to determine whether a reload
+// is required from systemd to take the new systemd files into effect. This is a NOP
+// if opts.Preseeding is set
+func (es ensureSnapServicesContext) reloadModified() error {
+	if es.opts.Preseeding {
+		return nil
+	}
 
+	if es.modifiedSystemServices {
+		if err := es.sysd.DaemonReload(); err != nil {
+			return err
+		}
+	}
+	if es.modifiedUserServices {
+		if err := userDaemonReload(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ensureSnapAppsSystemdServices takes care of writing .service files for all services
+// registered in snap.Info apps.
+func ensureSnapAppsSystemdServices(context *ensureSnapServicesContext, snapInfo *snap.Info, opts *AddSnapServicesOptions) error {
```

Comment:
I would go a little forward and make this a member function of the context:

```suggestion
func (es *ensureSnapServicesContext) ensureSnapAppsSystemdServices(snapInfo *snap.Info, opts *AddSnapServicesOptions) error {
```

And I'm not sure why this function has an "Apps" in its name -- I'd remove it. Also, since we only support systemd for services, the "Systemd" part could also be removed. My favourite would be to call this "ensureSnapSystemdUnits", anyway :-)

Reply:
Ah nice, good suggestion! Had a hard time coming up with namings!